File: samples/openssl-aes.py
#!/usr/bin/python2
"""
Copyright 2013 Christophe-Marie Duquesne <chmd@chmd.fr>

This file is in PUBLIC DOMAIN.

It contains function to Encrypt/Decrypt text in an openssl compatible way,
with minimal dependencies (pycrypto). See the functions aes_encrypt/aes_decrypt.
"""

import sys
import select
from base64 import b64decode, b64encode
from binascii import hexlify
from Crypto.Cipher import AES
from Crypto.Hash import MD5
# (!) We use a crypto-secure randint function
from Crypto.Random.random import randint

def random_salt(n=8):
    """Generate an 8 random bytes salt"""
    res = ''
    for i in range(n):
        res += chr(randint(0,255))
    return res

def derive_key_and_iv(secret, salt):
    """Derives a key and an iv from the password and the salt, using
    opensssl non standard method"""
    hash_1 = MD5.new(secret+salt).digest()
    hash_2 = MD5.new(hash_1+secret+salt).digest()
    hash_3 = MD5.new(hash_2+secret+salt).digest()
    mat = hash_1 + hash_2 + hash_3

    key = mat[0:32]
    iv  = mat[32:48]
    return (iv, key)

def aes_decrypt(secret, encrypted_text):
    """Decrypt the text. Does the same thing as:
    echo <encrypted_text> | openssl aes-256-cbc -d -base64 -pass pass:<secret>
    """
    encrypted = b64decode(encrypted_text)
    salt = encrypted[8:16]
    data = encrypted[16:]
    iv, key = derive_key_and_iv(secret, salt)
    cypher = AES.new(key, AES.MODE_CBC, iv)
    text = cypher.decrypt(data)
    return text

def aes_encrypt(secret, text):
    """Encrypt the text. Does the same thing as:
    echo <encrypted_text> | openssl aes-256-cbc -d -base64 -pass pass:<secret>
    """
    salt = random_salt()
    iv, key = derive_key_and_iv(secret, salt)
    cypher = AES.new(key, AES.MODE_CBC, iv)
    # PKCS#7: padd with n, where n is the number of characters remaining
    # to reach a multiple of 16.
    n = 16 - len(text) % 16
    text += chr(n) * n
    encrypted = cypher.encrypt(text)
    return b64encode('Salted__' + salt + encrypted)

def main():
    if (not select.select([sys.stdin,],[],[],0.0)[0] or len(sys.argv)!= 3
            or sys.argv[1] not in ("encrypt", "decrypt")):
        print "Usage: echo text | ./aes.py [decrypt|decrypt] <password>"
        exit(2)
    # Get the input data from stdin
    datain = ""
    for line in sys.stdin:
        datain += line
    # Get the password
    password = sys.argv[2]
    # Encrypt/decrypt
    if sys.argv[1] == "encrypt":
        sys.stdout.write(aes_encrypt(password, datain))
        sys.stdout.write("\n")
    if sys.argv[1] == "decrypt":
        sys.stdout.write(aes_decrypt(password, datain))

if __name__ == "__main__":
    main()

File: samples/OpensslAES.java
/*
 * Copyright 2013 Christophe-Marie Duquesne <chmd@chmd.fr>
 *
 * This file is in PUBLIC DOMAIN.
 *
 * It contains function to Encrypt/Decrypt text in an openssl compatible
 * way, with minimal dependencies (everything standard except Base64). See
 * the functions aesEncrypt and aesDecrypt.
 *
 * The Base64 class used here is stolen from android. You may find a copy
 * of it on http://minibackup.chmd.fr/samples/Base64.java
 */

import java.security.DigestException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class OpensslAES {

    /*
     * Derive an key and IV from a password and a salt
     * using OpenSSL's non-standard method.
     */
    static byte [] deriveKeyAndIV(byte[] password, byte[] salt) throws
        DigestException, NoSuchAlgorithmException {

        byte[] res = new byte[48];

        final MessageDigest md5 = MessageDigest.getInstance("MD5");

        md5.update(password);
        md5.update(salt);
        byte[] hash1 = md5.digest();

        md5.reset();
        md5.update(hash1);
        md5.update(password);
        md5.update(salt);
        byte[] hash2 = md5.digest();

        md5.reset();
        md5.update(hash2);
        md5.update(password);
        md5.update(salt);
        byte[] hash3 = md5.digest();

        // copy the 3 hashes in the result array
        System.arraycopy(hash1, 0, res, 0, 16);
        System.arraycopy(hash2, 0, res, 16, 16);
        System.arraycopy(hash3, 0, res, 32, 16);
        return res;
    }

    /*
     * Decrypt the string data, using the password secret,
     * in an openssl-compatible way. This function provides the equivalent
     * of the command:
     *
     * echo <data> | openssl aes-256-cbc -d -base64 -pass pass:<secret>
     */
    public static String aesDecrypt(String secret, String data)
            throws DigestException, NoSuchAlgorithmException,
                NoSuchPaddingException, InvalidKeyException,
                InvalidAlgorithmParameterException,
                IllegalBlockSizeException,
                BadPaddingException {

        byte[] decoded = Base64.decode(data, Base64.DEFAULT);
        byte[] salt = Arrays.copyOfRange(decoded, 8, 16);
        byte[] encrypted = Arrays.copyOfRange(decoded, 16, decoded.length);

        byte[] keyAndIV = deriveKeyAndIV(secret.getBytes(), salt);
        byte[] key = Arrays.copyOfRange(keyAndIV, 0, 32);
        byte[] iv = Arrays.copyOfRange(keyAndIV, 32, 48);

        SecretKeySpec skeySpec = new SecretKeySpec(key, "AES");
        IvParameterSpec ivspec = new IvParameterSpec(iv);

        Cipher cipher;
        cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, skeySpec, ivspec);
        return new String(cipher.doFinal(encrypted));
    }

    /*
     * Encrypt the string data, using the password secret.
     * This function provides the equivalent of the command:
     *
     * echo <data> | openssl aes-256-cbc -base64 -pass pass:<secret>
     */
    public static String aesEncrypt(String secret, String data)
            throws DigestException, NoSuchAlgorithmException,
                NoSuchPaddingException, InvalidKeyException,
                InvalidAlgorithmParameterException,
                IllegalBlockSizeException, BadPaddingException {
        byte [] salt = new byte [8];
        SecureRandom sr = new SecureRandom();
        sr.nextBytes(salt);

        byte [] keyAndIV = deriveKeyAndIV(secret.getBytes(), salt);
        byte [] key = Arrays.copyOfRange(keyAndIV, 0, 32);
        byte [] iv = Arrays.copyOfRange(keyAndIV, 32, 48);
        SecretKeySpec skeySpec = new SecretKeySpec(key, "AES");
        IvParameterSpec ivspec = new IvParameterSpec(iv);

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, skeySpec, ivspec);
        byte [] encrypted = cipher.doFinal(data.getBytes());

        byte [] to_encode = new byte[ 16 + encrypted.length ];
        System.arraycopy("Salted__".getBytes(), 0, to_encode, 0, 8);
        System.arraycopy(salt, 0, to_encode, 8, 8);
        System.arraycopy(encrypted, 0, to_encode, 16, encrypted.length);
        return Base64.encodeToString(to_encode, Base64.DEFAULT);
    }

    public static void main(String args[]) throws DigestException, NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException{

        // obtained with:
        // echo -n "original string" | openssl aes-256-cbc -base64 -pass pass:secret
        String enc = "U2FsdGVkX19VsjpO6ajerdwd8IfRTYMjs6oYOjHRepRoZWtfXvbKRZOgoR+GncsO";
        System.out.println(aesDecrypt("secret", enc));
        System.out.println(aesEncrypt("secret", "original string"));

    }
}
